kv: flake in TestDelegateSnapshot

Found here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/8578471?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

```
=== RUN   TestDelegateSnapshot
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/09357cecfdbbab5926b4b055936a9b62/logTestDelegateSnapshot725553609
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:410: could not add [n4,s4] replicas to range containing /Table/Max: AdminChangeReplicas error: change replicas of r59 failed: replica unavailable: (n1,s1):1 unable to serve request to r59:/{Table/Max-Max} [(n1,s1):1, (n3,s3):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":79,"next":80,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}: encountered poisoned latch /Local/Range/Table/Max/RangeDescriptor@0,0
        (1) AdminChangeReplicas error: change replicas of r59 failed: replica unavailable: (n1,s1):1 unable to serve request to r59:/{Table/Max-Max} [(n1,s1):1, (n3,s3):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":79,"next":80,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}: encountered poisoned latch /Local/Range/Table/Max/RangeDescriptor@0,0
          | -- cause hidden behind barrier
          | AdminChangeReplicas error: change replicas of r59 failed: replica unavailable: (n1,s1):1 unable to serve request to r59:/{Table/Max-Max} [(n1,s1):1, (n3,s3):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":79,"next":80,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}: encountered poisoned latch /Local/Range/Table/Max/RangeDescriptor@0,0
          | (1) attached stack trace
          |   -- stack trace:
          |   | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).changeReplicas
          |   |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:762
          |   | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).addReplica
          |   |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:772
          |   | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddVoters
          |   |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:790
          |   | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddVotersOrFatal
          |   |   github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:906
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestDelegateSnapshot
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:410
          |   | testing.tRunner
          |   |   GOROOT/src/testing/testing.go:1446
          |   | runtime.goexit
          |   |   GOROOT/src/runtime/asm_amd64.s:1594
          | Wraps: (2) AdminChangeReplicas error
          | Wraps: (3)
          |   | (opaque error wrapper)
          |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          |   | reportable 0:
          |   |
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.execChangeReplicasTxn
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:2451
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).initializeRaftLearners
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1751
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).changeReplicasImpl
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1065
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).ChangeReplicas
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:991
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeAdminBatch
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:957
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:187
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:205
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:200
          |   | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          |   |   github.com/cockroachdb/cockroach/pkg/server/node.go:1154
          |   | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          |   |   github.com/cockroachdb/cockroach/pkg/server/node.go:1226
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:762
          |   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
          |   |   github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:95
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:264
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:92
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:232
          |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          |   |   github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:322
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:230
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:772
          |   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ClientInterceptor.func2
          |   |   github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:226
          |   | github.com/cockroachdb/cockroach/pkg/rpc.getChainUnaryInvoker.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:967
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func3
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:841
          |   | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:975
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:209
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:188
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:2142
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1668
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1240
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:861
          | Wraps: (4) change replicas of r59 failed
          | Wraps: (5) replica unavailable: (n1,s1):1 unable to serve request to r59:/{Table/Max-Max} [(n1,s1):1, (n3,s3):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":79,"next":80,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}: encountered poisoned latch /Local/Range/Table/Max/RangeDescriptor@0,0
          | Wraps: (6)
          |   | (opaque error wrapper)
          |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          |   | reportable 0:
          |   |
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.replicaUnavailableError
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:276
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).replicaUnavailableError
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:285
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:478
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:180
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:205
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:200
          |   | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          |   |   github.com/cockroachdb/cockroach/pkg/server/node.go:1154
          |   | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          |   |   github.com/cockroachdb/cockroach/pkg/server/node.go:1226
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:762
          |   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
          |   |   github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:95
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:264
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:92
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:232
          |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          |   |   github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:322
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:230
          |   | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:883
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:772
          |   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ClientInterceptor.func2
          |   |   github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:226
          |   | github.com/cockroachdb/cockroach/pkg/rpc.getChainUnaryInvoker.func1
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:967
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func3
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:841
          |   | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
          |   |   github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:975
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:209
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:188
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:2142
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1668
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1437
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:861
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnLockGatekeeper).SendLocked
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_lock_gatekeeper.go:82
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnMetricRecorder).SendLocked
          |   |   github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_metric_recorder.go:46
          | Wraps: (7) closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":79,"next":80,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}
          | Wraps: (8) secondary error attachment
          |   | replica unavailable: (n1,s1):1 unable to serve request to r59:/{Table/Max-Max} [(n1,s1):1, (n3,s3):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":76,"next":77,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}: have been waiting 61.00s for slow proposal HeartbeatTxn [/Local/Range/Table/Max/RangeDescriptor,/Min), [txn: dd6c506d]
          |   | (1) forced error mark
          |   |   | "originated at breaker breaker"
          |   |   | github.com/cockroachdb/cockroach/pkg/util/circuit/*circuit.breakerErrorMark::
          |   | Wraps: (2) forced error mark
          |   |   | "breaker open"
          |   |   | github.com/cockroachdb/errors/withstack/*withstack.withStack::
          |   | Wraps: (3) replica unavailable: (n1,s1):1 unable to serve request to r59:/{Table/Max-Max} [(n1,s1):1, (n3,s3):2, next=3, gen=3, sticky=9223372036.854775807,2147483647]: closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":76,"next":77,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}: have been waiting 61.00s for slow proposal HeartbeatTxn [/Local/Range/Table/Max/RangeDescriptor,/Min), [txn: dd6c506d]
          |   | Wraps: (4)
          |   |   | (opaque error wrapper)
          |   |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          |   |   | reportable 0:
          |   |   |
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.replicaUnavailableError
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:276
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).replicaUnavailableError
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:285
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaCircuitBreaker).tripSync
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:109
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaCircuitBreaker).TripAsync.func1
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:103
          |   |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
          |   |   |   github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
          |   |   | runtime.goexit
          |   |   |   GOROOT/src/runtime/asm_amd64.s:1594
          |   | Wraps: (5) closed timestamp: 1675460129.386505563,0 (2023-02-03 21:35:29); raft status: {"id":"1","term":6,"vote":"1","commit":15,"lead":"1","raftState":"StateLeader","applied":15,"progress":{"1":{"match":76,"next":77,"state":"StateReplicate"},"2":{"match":13,"next":14,"state":"StateSnapshot"}},"leadtransferee":"0"}
          |   | Wraps: (6)
          |   |   | (opaque error wrapper)
          |   |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          |   |   | reportable 0:
          |   |   |
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).refreshProposalsLocked
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1366
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).tick
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:1229
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processTick
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:672
          |   |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker
          |   |   |   github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:327
          |   |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
          |   |   |   github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
          |   |   | runtime.goexit
          |   |   |   GOROOT/src/runtime/asm_amd64.s:1594
          |   | Wraps: (7) have been waiting 61.00s for slow proposal HeartbeatTxn [/Local/Range/Table/Max/RangeDescriptor,/Min), [txn: dd6c506d]
          |   | Error types: (1) *markers.withMark (2) *markers.withMark (3) *roachpb.ReplicaUnavailableError (4) *errbase.opaqueWrapper (5) *errutil.withPrefix (6) *errbase.opaqueWrapper (7) *errutil.leafError
          | Wraps: (9) forced error mark
          |   | "breaker open"
          |   | github.com/cockroachdb/errors/withstack/*withstack.withStack::
          | Wraps: (10) encountered poisoned latch /Local/Range/Table/Max/RangeDescriptor@0,0
          | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errbase.opaqueWrapper (4) *errutil.withPrefix (5) *roachpb.ReplicaUnavailableError (6) *errbase.opaqueWrapper (7) *errutil.withPrefix (8) *secondary.withSecondaryError (9) *markers.withMark (10) *poison.PoisonedError
        Error types: (1) *barriers.barrierErr
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/09357cecfdbbab5926b4b055936a9b62/logTestDelegateSnapshot725553609
--- FAIL: TestDelegateSnapshot (64.39s)
```

cc @andrewbaptist for triage

Jira issue: CRDB-24166

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -23,6 +23,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
@@ -352,74 +353,82 @@ func TestAddReplicaWithReceiverThrottling(t *testing.T) {
 
 // TestDelegateSnapshot verifies that the correct delegate is chosen when
 // sending snapshots to stores.
-// TODO: It is currently disabled because with raft snapshots sometimes being
-// required and not going through snapshot delegation, the results are
-// unpredictable.
 func TestDelegateSnapshot(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.WithIssue(t, 96841, "Occasionally fails until 87553 is resolved")
-
 	ctx := context.Background()
 
-	// Synchronize on the moment before the snapshot gets sent to measure the
-	// state at that time.
-	requestChannel := make(chan *kvserverpb.DelegateSendSnapshotRequest, 10)
-
-	setupFn := func(t *testing.T) (
-		*testcluster.TestCluster,
-		roachpb.Key,
-	) {
-		knobs, ltk := makeReplicationTestKnobs()
-		ltk.storeKnobs.DisableRaftSnapshotQueue = true
-
-		ltk.storeKnobs.SendSnapshot = func(request *kvserverpb.DelegateSendSnapshotRequest) {
+	// Record snapshots as they are sent on this channel for later analysis.
+	requestChannel := make(chan *kvserverpb.DelegateSendSnapshotRequest, 2)
+	knobs, ltk := makeReplicationTestKnobs()
+	ltk.storeKnobs.SendSnapshot = func(request *kvserverpb.DelegateSendSnapshotRequest) {
+		// TODO(abaptist): Remove this condition once 96841 is fixed. This
+		// accounts spurious raft snapshots that are sent. Also disable the raft
+		// snapshot queue using ltk.storeKnobs.DisableRaftSnapshotQueue = true.
+		if request.SenderQueueName != kvserverpb.SnapshotRequest_RAFT_SNAPSHOT_QUEUE {
 			requestChannel <- request
 		}
-
-		localityA := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "a"}}}
-		localityB := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "b"}}}
-
-		localityServerArgs := make(map[int]base.TestServerArgs)
-		localityServerArgs[0] = base.TestServerArgs{Knobs: knobs, Locality: localityA}
-		localityServerArgs[1] = base.TestServerArgs{Knobs: knobs, Locality: localityA}
-		localityServerArgs[2] = base.TestServerArgs{Knobs: knobs, Locality: localityB}
-		localityServerArgs[3] = base.TestServerArgs{Knobs: knobs, Locality: localityB}
-
-		tc := testcluster.StartTestCluster(
-			t, 4, base.TestClusterArgs{
-				ServerArgsPerNode: localityServerArgs,
-				ReplicationMode:   base.ReplicationManual,
-			},
-		)
-		scratchKey := tc.ScratchRange(t)
-		return tc, scratchKey
 	}
 
-	tc, scratchKey := setupFn(t)
+	localityA := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "a"}}}
+	localityB := roachpb.Locality{Tiers: []roachpb.Tier{{Key: "region", Value: "b"}}}
+	localityServerArgs := map[int]base.TestServerArgs{
+		0: {Knobs: knobs, Locality: localityA},
+		1: {Knobs: knobs, Locality: localityA},
+		2: {Knobs: knobs, Locality: localityB},
+		3: {Knobs: knobs, Locality: localityB},
+	}
+
+	tc := testcluster.StartTestCluster(
+		t, 4, base.TestClusterArgs{
+			ServerArgsPerNode: localityServerArgs,
+			ReplicationMode:   base.ReplicationManual,
+		},
+	)
+
+	scratchKey := tc.ScratchRange(t)
 	defer tc.Stopper().Stop(ctx)
-
-	// Node 3 (loc B) can only get the data from node 1 as its the only one that has it.
-	_ = tc.AddVotersOrFatal(t, scratchKey, tc.Targets(2)...)
-	request := <-requestChannel
-	require.Equal(t, request.DelegatedSender.StoreID, roachpb.StoreID(1))
-	// Drain the channel. Unfortunately there are occasionally spurious raft snapshots sent.
-	for len(requestChannel) > 0 {
-		<-requestChannel
+	// Node 3 (loc B) can only get the data from node 1 as it is the only replica.
+	{
+		_ = tc.AddVotersOrFatal(t, scratchKey, tc.Targets(2)...)
+		request := <-requestChannel
+		require.Equalf(t, request.DelegatedSender.StoreID, roachpb.StoreID(1), "Wrong sender for request %+v", request)
 	}
 
-	// Node 4 (loc B) should get the snapshot from node 3 as its the same locality.
-	_ = tc.AddVotersOrFatal(t, scratchKey, tc.Targets(3)...)
-	request = <-requestChannel
-	require.Equal(t, request.DelegatedSender.StoreID, roachpb.StoreID(3))
-	for len(requestChannel) > 0 {
-		<-requestChannel
+	// Node 4 (loc B) should get the delegated snapshot from node 3 which is the
+	// same locality.
+	{
+		leaderDesc := tc.AddVotersOrFatal(t, scratchKey, tc.Targets(3)...)
+		// Wait until we are sure the delegate store has received the descriptor. It
+		// can't delegate until it receives the latest generation descriptor.
+		testutils.SucceedsSoon(t, func() error {
+			var desc roachpb.RangeDescriptor
+			rKey := keys.MustAddr(scratchKey)
+			require.NoError(t, tc.Servers[2].DB().GetProto(ctx, keys.RangeDescriptorKey(rKey), &desc))
+			if desc.Generation != leaderDesc.Generation {
+				return errors.Newf("Generation mismatch %d != %d", desc.Generation, leaderDesc.Generation)
+			}
+			return nil
+		})
+		request := <-requestChannel
+		require.Equalf(t, request.DelegatedSender.StoreID, roachpb.StoreID(3), "Wrong type of request %+v", request)
+		// TODO(abaptist): Remove this loop. Sometimes the delegated request fails
+		// due to Raft updating the generation before we can delegate. Even with the
+		// loop above to get the delegate on the correct generation, this is racy if
+		// there is a raft snapshot. We fall back to not using our snapshot if the
+		// generation fails, but this means that a second request is sent from the
+		// leaseholder.
+		for len(requestChannel) > 0 {
+			<-requestChannel
+		}
 	}
 
-	// Node 2 (loc A) should get the snapshot from node 1 as it is the same locality.
-	_ = tc.AddVotersOrFatal(t, scratchKey, tc.Targets(1)...)
-	request = <-requestChannel
-	require.Equal(t, request.DelegatedSender.StoreID, roachpb.StoreID(1))
+	// Node 2 (loc A) should get the snapshot from node 1 as they have the same locality.
+	{
+		_ = tc.AddVotersOrFatal(t, scratchKey, tc.Targets(1)...)
+		request := <-requestChannel
+		require.Equalf(t, request.DelegatedSender.StoreID, roachpb.StoreID(1), "Wrong type of request %+v", request)
+	}
 }
 
 // TestDelegateSnapshotFails is a test that ensure we fail fast when the

